common/djangoapps/static_replace/services.py

```diff
@@ -0,0 +1,46 @@
+"""
+Supports replacement of static urls to absolute urls in xblocks.
+"""
+
+from xblock.reference.plugins import Service
+
+from common.djangoapps.static_replace import replace_course_urls, replace_jump_to_id_urls, replace_static_urls
+
+
+class ReplaceURLService(Service):
```

Comment:
Given that we're adding this service, should we also deprecate [these methods](https://github.com/openedx/edx-platform/blob/25b275bca4aee8099ad015648efa8551b272e11a/openedx/core/djangoapps/xblock/runtime/shims.py#L179-L219)?
cc: @bradenmacdonald, as I'm not sure if these aren't used by the blockstore.

Reply:
They're not called by blockstore but they are called by XBlocks, so yes.

@kaustavb12 Please make the same changes to https://github.com/openedx/edx-platform/blob/25b275bca4aee8099ad015648efa8551b272e11a/openedx/core/djangoapps/xblock/runtime/shims.py#L179-L219 - make them print a deprecation warning and call the new service instead.

Reply:
@bradenmacdonald @Agrendalath 

[These methods](https://github.com/openedx/edx-platform/blob/25b275bca4aee8099ad015648efa8551b272e11a/openedx/core/djangoapps/xblock/runtime/shims.py#L179-L219) are part of the [RuntimeShim class](https://github.com/openedx/edx-platform/blob/d44f021580f5ba12534a7035ab2b6f9c0245b588/openedx/core/djangoapps/xblock/runtime/shims.py#L20) which is used with the [XBlockRuntime class](https://github.com/openedx/edx-platform/blob/d44f021580f5ba12534a7035ab2b6f9c0245b588/openedx/core/djangoapps/xblock/runtime/runtime.py#L60).
The [XBlockRuntime class](https://github.com/openedx/edx-platform/blob/d44f021580f5ba12534a7035ab2b6f9c0245b588/openedx/core/djangoapps/xblock/runtime/runtime.py#L60) is a super class of the [BlockstoreXBlockRuntime class](https://github.com/openedx/edx-platform/blob/d44f021580f5ba12534a7035ab2b6f9c0245b588/openedx/core/djangoapps/xblock/runtime/blockstore_runtime.py#L28)

As I understand, blockstore Xblocks are not attached to courses, and so they don't have course id associated with them.

In that case would having [replace_course_urls()](https://github.com/openedx/edx-platform/blob/d44f021580f5ba12534a7035ab2b6f9c0245b588/openedx/core/djangoapps/xblock/runtime/shims.py#L200) and [replace_jump_to_id_urls()](https://github.com/openedx/edx-platform/blob/d44f021580f5ba12534a7035ab2b6f9c0245b588/openedx/core/djangoapps/xblock/runtime/shims.py#L212) methods in this shim necessary, since replacing both course urls and jump-to-id urls makes use of course ids.

Reply:
Let me know if you want more help!

Reply:
@bradenmacdonald, I'm copying the message from @kaustavb12 here.

> I have used the ReplaceURLService directly [here](https://github.com/open-craft/edx-platform/blob/c1c323ce08ac200955e921798039a014688bb264/openedx/core/djangoapps/xblock/runtime/runtime.py#L337) instead of adding the service in the [service() method](https://github.com/open-craft/edx-platform/blob/c1c323ce08ac200955e921798039a014688bb264/openedx/core/djangoapps/xblock/runtime/runtime.py#L212-L251). This is because, for invoking the service from this method we need to add a declaration for each Xblock in the form of @XBlock.needs('mako'). However, I couldn't figure out the correct classes for xblocks like Completion, Drag and Drop, Feedback, Survey, etc. 

> If you think the correct implementation should have been through the service() method, then I will look into in further later in the sprint.

As I don't know what this change means for blockstore, I could use a second pair of eyes to check the approach from c1c323ce08ac200955e921798039a014688bb264 when you have a moment. For now, we don't need to use the attributes of ReplaceURLService here, so we could make `ReplaceURLService.replace_block_urls` a static method to start using it as an API. If we are going to support courses with blockstore in the future, then we could expand the service to support them, and merge `replace_static_block_urls` with `replace_static_urls` within `static_replace/__init__.py`. Would it make sense?

Reply:
@Agrendalath Sure, I left my comments on c1c323ce08ac200955e921798039a014688bb264

Reply:
@bradenmacdonald 
Thanks a lot pointing me to the instructions to setup blockstore in devstack and for reviewing the changes. It really helped me gain a lot of context about xblocks and the blockstore runtime.

I have now incorporated all your suggestions, and now there is only one API in [ReplaceURLService](https://github.com/open-craft/edx-platform/blob/kaustav/bd_13_replace_urls/common/djangoapps/static_replace/services.py) which can be called by XBlocks as needed in the form of `self.runtime.service(self, 'replace_urls').replace_urls(text)`

Reply:
@bradenmacdonald 

Also I have gone ahead and merged the new `replace_static_block_urls` function with the `replace_static_urls` function in [static_replace/__init__.py](https://github.com/open-craft/edx-platform/blob/8fb0a90039ae58fd686b8e550ce6d2aa98974240/common/djangoapps/static_replace/__init__.py#L150-L192) similar to what was done for [ReplaceURLService](https://github.com/open-craft/edx-platform/blob/kaustav/bd_13_replace_urls/common/djangoapps/static_replace/services.py) to reduce code duplication.

Reply:
Thanks @kaustavb12, those changes look great :)